trino-aws-proxy/src/test/java/io/trino/aws/proxy/server/rest/TestAwsChunkedInputStream.java

```diff
@@ -0,0 +1,337 @@
+            = "123456789012345612345";
+
+    @Test
+    public void testChunkedInputStreamLargeBuffer()
+            throws IOException
+    {
+        ByteArrayInputStream inputStream = new ByteArrayInputStream(CHUNKED_INPUT.getBytes(UTF_8));
+        InputStream in = new AwsChunkedInputStream(inputStream, new DummyChunkSigningSession());
+        byte[] buffer = new byte[300];
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        int len;
+        while ((len = in.read(buffer)) > 0) {
+            out.write(buffer, 0, len);
+        }
+        assertEquals(-1, in.read(buffer));
+        assertEquals(-1, in.read(buffer));
+
+        in.close();
+
+        String result = out.toString(UTF_8);
+        assertEquals(CHUNKED_RESULT, result);
+    }
+
+    //Test for when buffer is smaller than chunk size.
+    @Test
+    public void testChunkedInputStreamSmallBuffer()
+            throws IOException
+    {
+        ByteArrayInputStream inputStream = new ByteArrayInputStream(CHUNKED_INPUT.getBytes(StandardCharsets.US_ASCII));
+        InputStream in = new AwsChunkedInputStream(inputStream, new DummyChunkSigningSession());
+
+        byte[] buffer = new byte[7];
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        int len;
+        while ((len = in.read(buffer)) > 0) {
+            out.write(buffer, 0, len);
+        }
+        assertEquals(-1, in.read(buffer));
+        assertEquals(-1, in.read(buffer));
+    }
+
+    // One byte read
+    @Test
+    public void testChunkedInputStreamOneByteRead()
+            throws IOException
+    {
+        String s = "5;chunk-signature=0\r\n01234\r\n5;chunk-signature=0\r\n56789\r\n0;chunk-signature=0\r\n\r\n";
+        ByteArrayInputStream inputStream = new ByteArrayInputStream(s.getBytes(StandardCharsets.US_ASCII));
+        InputStream in = new AwsChunkedInputStream(inputStream, new DummyChunkSigningSession());
+        int ch;
+        int i = '0';
+        while ((ch = in.read()) != -1) {
+            assertEquals(i, ch);
+            i++;
+        }
+        assertEquals(-1, in.read());
+        assertEquals(-1, in.read());
+
+        in.close();
+    }
+
+    // Missing closing chunk
+    @Test
+    public void testChunkedInputStreamNoClosingChunk()
+    {
+        String s = "5;chunk-signature=0\r\n01234\r\n";
+        ByteArrayInputStream inputStream = new ByteArrayInputStream(s.getBytes(StandardCharsets.US_ASCII));
+        InputStream in = new AwsChunkedInputStream(inputStream, new DummyChunkSigningSession());
+        byte[] tmp = new byte[5];
+        // altered from original test. Our AwsChunkedInputStream is improved and throws when the final chunk is missing or bad
+        assertThrows(IOException.class, () -> in.read(tmp));
+    }
+
+    // Truncated stream (missing closing CRLF)
+    @Test
+    public void testCorruptChunkedInputStreamTruncatedCRLF()
+            throws IOException
+    {
+        String s = "5;chunk-signature=0\r\n01234";
```

Comment:
This is a good test, but I'd suggest checking:
- `5;chunk-signature=0\r\n01234` (the current one): valid chunk without either a CRLF nor a final chunk
- `5;chunk-signature=0\r\n01234\r\n`: valid chunk without a final chunk
- `5;chunk-signature=0\r\n012340;chunk-signature=0\r\n\r\n`: valid chunk without a CRLF between it and the next (Final) chunk